integration/experiment/frequency_sweep/gen_policy_recommendation.py

```diff
@@ -0,0 +1,481 @@
+    if region_filter and region_filter != 'Epoch':
+        df_filtered = df[df['region'].isin(region_filter.split(','))]
+
+    # these are the only columns we need
+    df_cols = df_filtered[['FREQ_DEFAULT',
+                           'host', 'Profile',
+                           'runtime (sec)',
+                           'package-energy (joules)']]
+
+    df_cols = df_cols.rename({'FREQ_DEFAULT': 'freq',
+                              'runtime (sec)': 'runtime',
+                              'package-energy (joules)': 'energy'
+                             }, axis = 1)
+
+    df_cols = df_cols.groupby(['freq', 'host', 'Profile']).sum().reset_index()
+    df_cols = df_cols[['freq', 'runtime', 'energy']].set_index('freq')
+
+    return df_cols
+
+
+def dump_stats_summary(df, fname):
+    """
+    Write mean runtime and energy and the standard deviation of
+    runtime and energy for each frequency in CSV format to the
+    file fname."""
+    means = df.groupby(level=0).mean().rename(lambda x: x+"_mean", axis=1)
+    stds = df.groupby(level=0).std().rename(lambda x: x+"_stddev", axis=1)
+    pandas.concat([means, stds], axis=1).to_csv(fname)
+
+
+def policy_min_energy(freqrange, enmodel, rtmodel = None, sticker = None,
+                      max_degradation = None):
+    """
+    Find the frequency over the range freqrange (list-like) that
+    has the minimum predicted energy usage (according to the energy model
+    enmodel, an instance of PowerLimitModel), subject to the constraint
+    that its runtime does not exceed the runtime at sticker frequency
+    by more than a factor of (1 + max_degradation), if max_degradation is
+    specified. Returns a dictionary with keys power, runtime, and energy,
+    and values the optimal frequency and the predicted runtime and energy
+    at that limit, respectively."""
+
+    if sticker is None:
+        sticker = max(freqrange)
+
+    en_predictions = enmodel.batch_evaluate(freqrange)
+    if max_degradation is None:
+        # we don't need the runtime model in this case
+        best_energy, best_freq = min(zip(en_predictions, freqrange))
+        if rtmodel:
+            best_runtime = rtmodel.evaluate(best_freq)
+        else:
+            best_runtime = None
+        return {'freq': best_freq,
+                'runtime': best_runtime,
+                'energy': best_energy}
+    else:
+        rt_predictions = rtmodel.batch_evaluate(freqrange)
+        rt_at_sticker = rtmodel.evaluate(sticker)
+        constrained_values = [(energy, runtime, freq)
+                              for freq, runtime, energy
+                              in zip(freqrange, rt_predictions, en_predictions)
+                              if runtime <= (1 + max_degradation) * rt_at_sticker]
+        best_energy, best_runtime, best_freq = min(constrained_values)
+        return {'freq': best_freq,
+                'runtime': best_runtime,
+                'energy': best_energy}
+
+
+def rootssq(ll):
```

Comment:
These are also used by the power analysis scripts.  Please do not copy/paste implementations, rather include a common module.  

https://github.com/geopm/geopm/blob/dev/integration/experiment/power_sweep/gen_policy_recommendation.py#L260-L282